core_v14/src/decoder/decoder.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+// This file is part of substrate-desub.
+//
+// substrate-desub is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// substrate-desub is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with substrate-desub.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::decode_type::{decode_type, decode_type_by_id, DecodeTypeError};
+use super::extrinsic_bytes::{ExtrinsicBytes, ExtrinsicBytesError};
+use crate::metadata::{Metadata};
+use crate::substrate_value::SubstrateValue;
+use sp_runtime::{ MultiAddress, MultiSignature, AccountId32 };
+use codec::{ Decode };
+
+pub struct Decoder {
+	metadata: Metadata,
+}
+
+#[derive(Clone, Debug, thiserror::Error)]
+pub enum DecodeError<'a> {
+	#[error("Failed to parse the provided vector of extrinsics: {0}")]
+	UnexpectedExtrinsicsShape(#[from] ExtrinsicBytesError),
+    #[error("Failed to decode: {0}")]
+    CodecError(#[from] codec::Error),
+	#[error("Failed to decode type: {0}")]
+	DecodeTypeError(#[from] DecodeTypeError),
+	#[error("Failed to decode: expected more data")]
+	EarlyEof(&'static str),
+	#[error("Failed to decode unsupported extrinsic version '{0}'")]
+	CannotDecodeExtrinsicVersion(u8),
+	#[error("Cannot find call corresponding to extrinsic with pallet index {0} and call index {1}")]
+	CannotFindCall(u8, u8),
+	#[error("Failed to decode extrinsic: cannot find type ID {0}")]
+	CannotFindType(u32),
+	#[error("Decoding an extrinsic should consume all bytes, but {} bytes remain", .0.len())]
+	LateEof(&'a [u8], GenericExtrinsic)
```

Comment:
I don't think static lifetimes will be an issue, since the lifetime is based on the slice you pass in to the `decode` call, so eg:

```
let bytes: Vec<u8> = vec![1,2,3,4,5];
let err = decoder.decode_extrinsics(&data).unwrap_err();
// err should be happy now until bytes goes out of scope
```

That said, the lifetime isn't super ergonomic, so I'll turn it into a length offset, and probably map the error in `decode_extrinsics` so that it's a valid "number of bytes left" in either case!